### PR TITLE
[ENH] Add connectivity simulation function

### PR DESCRIPTION
**Pull request by tsbinns**
## Summary

Adds a new public function `make_signals_in_freq_bands()` inside `mne_connectivity/datasets/frequency.py` for a basic simulation of connectivity, following @adam2392's suggestion in #163.

<br>

## API/Documentation

The function has parameters for what I feel are the most important simulation options.

I added a new "Dataset functions" section to the API page since I wasn't sure how well it fit under the existing headers and link to the function here. Very open to change this if you have better suggestions.

<br>

## Unit tests

I added a new set of tests in `mne_connectivity/tests/test_datasets.py`.

`test_make_signals_error_catch()` checks that some obvious errors are caught. If you think there are some important checks I've missed please let me know.

`test_make_signals_in_freq_bands` checks that the simulations are actually working properly. I focused on Coh, ImCoh, and DPLI with: 1) no time delay in interaction between seeds and targets; 2) a positive time delay (i.e. seeds drive communication with targets); and 3) a negative time delay (i.e. targets drive communication with seeds). From this we expect:

- Case 1 (no time delay) - Coh should be high for the simulated frequency band, ImCoh should be low, DPLI should be ~0.5.
- Case 2 (positive delay) - Coh should be high, ImCoh should be high, DPLI should be > 0.5.
- Case 3 (negative delay) - Coh should be high, ImCoh should be high, DPLI should be < 0.5.

I test these with a higher and lower SNR, different number of seeds and targets, and the different CSD computation modes. I realise it's not an exhaustive check, but I think the tests nicely capture what we expect to see from undirected (Coh, ImCoh) and directed (DPLI) connectivity methods, and also methods that do not capture zero time-lag interactions (ImCoh).

<br>

## Use elsewhere in the package

This simulation function is similar (**but not identical**) to those used for many of the unit tests in `spectral.py`. Out of interest I  wanted to see if this function could be used to replace these simulations. In a number of cases it works, but in others it would require redefining what is considered acceptable connectivity values for particular method/CSD mode combinations.

This could be done, but it would require a considerable amount of effort tweaking these thresholds which I can't really invest the time for at the moment.

How would people feel about just having this new function which can be used going forward for e.g. the CaCoh examples in #163 without changing the existing unit test stimulations?

### CODE REVIEW

**Comment by larsoner** [comment]
> This could be done, but it would require a considerable amount of effort tweaking these thresholds which I can't really invest the time for at the moment

Can you just add a `# TODO: Use <whatever> for this after tweaking tolerances` so we don't forget that this is a possibility? Seems okay to leave it to future work

File: mne_connectivity/datasets/frequency.py [hunk 1]
```diff
@@ -0,0 +1,146 @@
...
+        Signal-to-noise ratio of the simulated data in the range [0, 1].
+    connection_delay : int (default 5)
+        Number of timepoints for the delay of connectivity between the seeds and
+        targets. If > 0, the target data is a delayed form of the seed data. If < 0, the
+        seed data is a delayed form of the target data.
+    tmin : int | float (default 0)
+        Earliest time of each epoch.
+    ch_names : list of str | None (default None)
+        Names of the channels in the simulated data. If `None`, the channels are named
+        according to their index and the frequency band of interaction.
```
Comment (adam2392) [inline mne_connectivity/datasets/frequency.py:57]:
```suggestion
        Names of the channels in the simulated data. If `None`, the channels are named
        according to their index and the frequency band of interaction. If specified, must be a list of
        ``n_seeds + n_targets`` channel names.
```

Is this correct?

File: mne_connectivity/datasets/frequency.py [hunk 2]
```diff
@@ -0,0 +1,146 @@
...
+    ch_types : str | list of str (default "eeg")
```
Comment (adam2392) [inline mne_connectivity/datasets/frequency.py:58]: Perhaps 

```
If specified as a list, must be a list of ``n_seeds + n_targets`` channel names.
```

**Comment by tsbinns** [comment]
@larsoner @adam2392 Thanks both for the feedback!

I replaced the simulations where no values were being checked (e.g. for error catching parameters, saving & loading, etc...), and marked 5 tests where the simulations could be replaced with `make_signals_in_freq_bands` following some tweaks to the acceptable values.

File: mne_connectivity/datasets/frequency.py [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+        frequency.
+    n_epochs : int (default 10)
+        Number of epochs in the simulated data.
+    n_times : int (default 200)
+        Number of timepoints each epoch of the simulated data.
+    sfreq : int | float (default 100)
+        Sampling frequency of the simulated data, in Hz.
+    trans_bandwidth : int | float (default 1)
+        Transition bandwidth of the filter to apply to isolate activity in
+        ``freq_band``, in Hz.
```
Comment (adam2392) [inline mne_connectivity/datasets/frequency.py:46]:
```suggestion
    trans_bandwidth : int | float (default 1)
        Transition bandwidth of the filter to apply to isolate activity in
        ``freq_band``, in Hz. These are passed to the ``l_bandwidth`` and ``h_bandwidth``
        keyword arguments in :func:`mne.filter.create_filter`.
```

File: mne_connectivity/datasets/frequency.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        sfreq=sfreq,
+        l_freq=freq_band[0],
+        h_freq=freq_band[1],
+        l_trans_bandwidth=trans_bandwidth,
+        h_trans_bandwidth=trans_bandwidth,
+        fir_design="firwin2",
+    )
+
+    # simulate noise for each channel
+    noise = rng.randn(n_channels, n_epochs * n_times + np.abs(connection_delay))
```
Comment (adam2392) [inline mne_connectivity/datasets/frequency.py:118]:
```suggestion
    noise = rng.standard_normal(size=(n_channels, n_epochs * n_times + np.abs(connection_delay)))
```

https://numpy.org/doc/stable/reference/random/generated/numpy.random.randn.html

File: mne_connectivity/datasets/frequency.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    if snr < 0 or snr > 1:
+        raise ValueError("Signal-to-noise ratio must be between 0 and 1.")
+
+    if np.abs(connection_delay) >= n_epochs * n_times:
+        raise ValueError(
+            "Connection delay must be less than the total number of timepoints."
+        )
+
+    # simulate data
+    rng = np.random.RandomState(rng_seed)
```
Comment (adam2392) [inline mne_connectivity/datasets/frequency.py:103]:
```suggestion
    rng = np.random.default_rng(rng_seed)
```

File: mne_connectivity/datasets/frequency.py [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+
+    # simulate signal source at desired frequency band
+    signal = rng.randn(1, n_epochs * n_times + np.abs(connection_delay))
```
Comment (adam2392) [inline mne_connectivity/datasets/frequency.py:106]:
```suggestion
    signal = rng.standard_normal(size=(1, n_epochs * n_times + np.abs(connection_delay)))
```

**Review by adam2392 — CHANGES_REQUESTED** [review]
Some minor changes, but this overall looks good to me

File: mne_connectivity/datasets/frequency.py (hunk 4 above)
Comment (tsbinns) [reply mne_connectivity/datasets/frequency.py:118]: Good to know!

**Comment by tsbinns** [comment]
Looks like some of the thresholds were just on the boundary and updating the way in which the random numbers were generated caused these to fail, so tweaked the thresholds slightly.

**Review by adam2392 — APPROVED** [review]
Nice!

**Comment by adam2392** [comment]
I'll let @larsoner and @drammock take a quick look.

Perhaps @ruuskas is also interested given he found a bug with the previous implementation? Hopefully this will simplify life for generating simulated data with expected correlations in the frequency domain

**Comment by adam2392** [comment]
Perhaps something to help move this forward is refactoring #163 to use this function? (I.e. merge changes might work?...)

I would keep a local copy in case git messes up. 

Then we can see the simulation function work as intended for the new PR?

If that's a bit hairy, I'm okay merging for now. Wdyt @larsoner

**Comment by tsbinns** [comment]
@adam2392 Sure, I'm open to this. Will be away at a conference rest of this week, but can start on Monday.

**Comment by adam2392** [comment]
No rush! Enjoy!

**Comment by tsbinns** [comment]
@adam2392 I've refactored #163 to use the new simulation function in the CaCoh and coherency method comparison examples where I was previously using a function within each example file.
Everything still works, no problems switching to the dedicated function.

**Comment by adam2392** [comment]
Ah very cool! I think this LGTM as a start to simplifying data generation in our connectivity analyses.

**Pull request merged**